File: DotNet/Messaging/MessageFetchers/HydraByTopicByDestinationMessageFetcher.cs
using Bollywell.Hydra.Messaging.MessageIds;
using LoveSeat;
using LoveSeat.Interfaces;

namespace Bollywell.Hydra.Messaging.MessageFetchers
{
    public class HydraByTopicByDestinationMessageFetcher : MessageFetcherBase<HydraMessage>
    {
        private readonly string _topic;
        private readonly string _destination;
        private readonly IKeyOptions _endKey;

        /// <summary>
        /// Initialise HydraByTopicByDestinationMessageFetcher
        /// </summary>
        /// <param name="topic">The topic for which to fetch messages</param>
        /// <param name="destination">The destination for which to fetch messages</param>
        public HydraByTopicByDestinationMessageFetcher(string topic, string destination)
        {
            _topic = topic;
            _destination = destination;
            _endKey = new KeyOptions(_topic, _destination, CouchValue.MaxValue);
        }

        protected override string ViewName { get { return "directedMessages"; } }

        // The directedMessages view is indexed on [topic, destination, id]
        protected override IKeyOptions MessageKey(IMessageId id) { return new KeyOptions(_topic, _destination, id.ToDocId()); }
        protected override IKeyOptions EndKey() { return _endKey; }

    }
}

File: Java/hydra/src/main/java/uk/co/shastra/hydra/messaging/messagefetchers/HydraByTopicByDestinationMessageFetcher.java
package uk.co.shastra.hydra.messaging.messagefetchers;

import org.ektorp.ComplexKey;

import uk.co.shastra.hydra.messaging.HydraMessage;
import uk.co.shastra.hydra.messaging.messageids.MessageId;

public class HydraByTopicByDestinationMessageFetcher extends MessageFetcherBase<HydraMessage> {

	private String topic;
	private String destination;
	private ComplexKey endKey;
    
	public HydraByTopicByDestinationMessageFetcher(String topic, String destination) {
		super(HydraMessage.class);
		this.topic = topic;
		this.destination = destination;
		// TODO - check that "{}" is a suitable max value
		endKey = ComplexKey.of(topic, destination, ComplexKey.emptyObject());
	}
	
	@Override
	protected String getViewName() { return "directedMessages"; }

    // The directedMessages view is indexed on [topic, destination, id]
	@Override
	protected ComplexKey getMessageKey(MessageId id) { return ComplexKey.of(topic, destination, id.toDocId()); }

	@Override
	protected ComplexKey getEndKey() { return endKey; }
	
}
